pkg/chunk/storage/caching_storage_client.go

```diff
@@ -0,0 +1,58 @@
+package storage
+
+import (
+	"context"
+	"time"
+
+	"github.com/weaveworks/cortex/pkg/chunk"
+)
+
+type cachingStorageClient struct {
+	chunk.StorageClient
+	cache    *chunk.FifoCache
+	validity time.Duration
+}
+
+func newCachingStorageClient(client chunk.StorageClient, size int, validity time.Duration) chunk.StorageClient {
+	if size == 0 {
+		return client
+	}
+
+	return &cachingStorageClient{
+		StorageClient: client,
+		cache:         chunk.NewFifoCache(size),
+		validity:      validity,
+	}
+}
+
+func (s *cachingStorageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch) (shouldContinue bool)) error {
+	if len(query.RangeValuePrefix) != 0 || len(query.RangeValueStart) != 0 { // Only caching queries that read full rows.
+		return s.StorageClient.QueryPages(ctx, query, callback)
+	}
+
+	value, updated, ok := s.cache.Get(query.TableName + query.HashValue)
+	if ok && time.Now().Sub(updated) < s.validity {
```

Comment:
We have similar logic in other places where we use the FIFOCache, might be better to move it there.